File: Cheater-iOS/Services/AIService.swift
//
//  AIService.swift
//  Cheater-iOS
//
//  AI service for quiz generation using Claude API
//

import Foundation
import UIKit

// MARK: - AI Errors
enum AIError: LocalizedError {
    case missingAPIKey
    case invalidRequest
    case networkError(Error)
    case invalidResponse
    case parsingError(String)
    case quotaExceeded
    case serverError(Int)

    var errorDescription: String? {
        switch self {
        case .missingAPIKey:
            return "API key not configured"
        case .invalidRequest:
            return "Invalid request parameters"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .invalidResponse:
            return "Invalid response from AI service"
        case .parsingError(let message):
            return "Failed to parse quiz: \(message)"
        case .quotaExceeded:
            return "API quota exceeded. Please try again later."
        case .serverError(let code):
            return "Server error (code: \(code))"
        }
    }

    var recoverySuggestion: String? {
        switch self {
        case .missingAPIKey:
            return "Please configure your Claude API key in Config.plist"
        case .networkError:
            return "Please check your internet connection and try again."
        case .quotaExceeded:
            return "You've reached your API limit. Please wait or upgrade your plan."
        default:
            return "Please try again. If the problem persists, contact support."
        }
    }
}

// MARK: - AI Service
actor AIService {
    // MARK: - Properties

    private let apiKey: String
    private let apiURL = URL(string: "https://api.anthropic.com/v1/messages")!
    private let model = "claude-sonnet-4-20250514"
    private let maxTokens = 4096

    // MARK: - Initialization

    init(apiKey: String? = nil) throws {
        // Try to get API key from parameter or Config.plist
        if let key = apiKey {
            self.apiKey = key
        } else if let configPath = Bundle.main.path(forResource: "Config", ofType: "plist"),
                  let config = NSDictionary(contentsOfFile: configPath),
                  let key = config["ANTHROPIC_API_KEY"] as? String,
                  !key.isEmpty,
                  key != "your-api-key-here" {
            self.apiKey = key
        } else {
            throw AIError.missingAPIKey
        }
    }

    // MARK: - Public Methods

    /// Generate quiz from OCR text
    /// - Parameters:
    ///   - text: Extracted text from homework
    ///   - subject: Optional subject hint
    /// - Returns: Generated Quiz
    /// - Throws: AIError if generation fails
    func generateQuiz(from text: String, subject: String? = nil) async throws -> Quiz {
        // Validate input
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw AIError.invalidRequest
        }

        // Build prompt
        let prompt = buildPrompt(text: text, subject: subject)

        // Create request
        var request = URLRequest(url: apiURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")

        // Build request body
        let requestBody: [String: Any] = [
            "model": model,
            "max_tokens": maxTokens,
            "messages": [
                [
                    "role": "user",
                    "content": prompt
                ]
            ]
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)

        // Make request
        let (data, response) = try await URLSession.shared.data(for: request)

        // Validate response
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200:
            break
        case 429:
            throw AIError.quotaExceeded
        case 400..<500:
            throw AIError.invalidRequest
        case 500..<600:
            throw AIError.serverError(httpResponse.statusCode)
        default:
            throw AIError.invalidResponse
        }

        // Parse response
        let quiz = try parseQuizResponse(data: data)

        return quiz
    }

    /// Generate quiz from homework image using Claude Vision with smart classification
    /// - Parameters:
    ///   - image: UIImage of homework
    ///   - subject: Optional subject hint
    /// - Returns: Generated Quiz
    /// - Throws: AIError if generation fails
    func generateQuiz(from image: UIImage, subject: String? = nil) async throws -> Quiz {
        print("🖼️ Using Claude Vision with smart classification to generate quiz")

        // Step 1: Perform local classification for topic hint
        let classifier = ContentClassifier()
        var detectedTopic = "generic"
        var classificationConfidence = 0.5

        do {
            let classification = try await classifier.classifyImage(image)
            detectedTopic = classification.topic
            classificationConfidence = classification.confidence

            print("🔍 Local classification: \(detectedTopic) (confidence: \(String(format: "%.2f", classificationConfidence)))")

            // Only use detected topic if confidence is above threshold
            if classificationConfidence < PromptManager.shared.classificationThreshold {
                print("⚠️ Classification confidence below threshold, using generic prompt")
                detectedTopic = "generic"
            }
        } catch {
            print("⚠️ Classification failed, using generic prompt: \(error.localizedDescription)")
            detectedTopic = "generic"
        }

        // Step 2: Convert image to base64
        let base64Image = try convertImageToBase64(image)

        // Step 3: Build topic-specific vision prompt
        let promptText = buildVisionPrompt(subject: subject, topic: detectedTopic)

        // Step 4: Create request
        var request = URLRequest(url: apiURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")

        // Build multimodal request body
        let requestBody: [String: Any] = [
            "model": model,
            "max_tokens": maxTokens,
            "messages": [
                [
                    "role": "user",
                    "content": [
                        [
                            "type": "image",
                            "source": [
                                "type": "base64",
                                "media_type": "image/jpeg",
                                "data": base64Image
                            ]
                        ],
                        [
                            "type": "text",
                            "text": promptText
                        ]
                    ]
                ]
            ]
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)

        // Step 5: Make request
        print("📡 Sending image to Claude Vision API with \(detectedTopic) prompt...")
        let (data, response) = try await URLSession.shared.data(for: request)

        // Validate response
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200:
            print("✅ Claude Vision API returned successfully")
            break
        case 429:
            throw AIError.quotaExceeded
        case 400..<500:
            throw AIError.invalidRequest
        case 500..<600:
            throw AIError.serverError(httpResponse.statusCode)
        default:
            throw AIError.invalidResponse
        }

        // Step 6: Parse response (now includes topic validation from Claude)
        let quiz = try parseQuizResponse(data: data)
        print("✅ Successfully generated \(quiz.questions.count) questions for topic: \(detectedTopic)")

        return quiz
    }

    // MARK: - Private Methods

    private func buildPrompt(text: String, subject: String?) -> String {
        return PromptManager.shared.buildTextPrompt(text: text, subject: subject)
    }

    private func parseQuizResponse(data: Data) throws -> Quiz {
        // Parse Claude API response
        struct ClaudeResponse: Codable {
            struct Content: Codable {
                let text: String
            }
            let content: [Content]
        }

        let decoder = JSONDecoder()

        // Decode outer Claude response
        guard let claudeResponse = try? decoder.decode(ClaudeResponse.self, from: data),
              let contentText = claudeResponse.content.first?.text else {
            throw AIError.parsingError("Failed to decode Claude response")
        }

        // Extract JSON from response (remove any markdown formatting)
        let jsonText = extractJSON(from: contentText)

        guard let jsonData = jsonText.data(using: .utf8) else {
            throw AIError.parsingError("Failed to convert text to data")
        }

        // Parse quiz JSON
        let quizResponse = try decoder.decode(Quiz.ClaudeResponse.self, from: jsonData)

        // Validate quiz
        guard quizResponse.questions.count == 10 else {
            throw AIError.parsingError("Quiz must have exactly 10 questions, got \(quizResponse.questions.count)")
        }

        // Validate all questions
        for (index, question) in quizResponse.questions.enumerated() {
            guard question.options.count == 4 else {
                throw AIError.parsingError("Question \(index + 1) must have 4 options, got \(question.options.count)")
            }
            guard question.correctIndex >= 0 && question.correctIndex < 4 else {
                throw AIError.parsingError("Question \(index + 1) has invalid correctIndex: \(question.correctIndex)")
            }
        }

        // Create Quiz (with placeholder homework ID)
        let quiz = Quiz(from: quizResponse, homeworkId: UUID())

        return quiz
    }

    private func extractJSON(from text: String) -> String {
        // Remove markdown code blocks if present
        var cleaned = text.trimmingCharacters(in: .whitespacesAndNewlines)

        // Remove ```json and ``` if present
        if cleaned.hasPrefix("```json") {
            cleaned = cleaned.replacingOccurrences(of: "```json", with: "")
        }
        if cleaned.hasPrefix("```") {
            cleaned = cleaned.replacingOccurrences(of: "```", with: "")
        }
        if cleaned.hasSuffix("```") {
            cleaned = String(cleaned.dropLast(3))
        }

        // Try to find JSON object boundaries
        if let startIndex = cleaned.firstIndex(of: "{"),
           let endIndex = cleaned.lastIndex(of: "}") {
            cleaned = String(cleaned[startIndex...endIndex])
        }

        return cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func buildVisionPrompt(subject: String?, topic: String = "generic") -> String {
        return PromptManager.shared.buildVisionPrompt(subject: subject, topic: topic)
    }

    private func convertImageToBase64(_ image: UIImage) throws -> String {
        // Resize image if too large to reduce API costs
        let maxDimension: CGFloat = 1568  // Claude recommended max
        let resizedImage = resizeImage(image, maxDimension: maxDimension)

        // Convert to JPEG with 80% quality for good balance of quality/size
        guard let imageData = resizedImage.jpegData(compressionQuality: 0.8) else {
            throw AIError.invalidRequest
        }

        print("📸 Image size: \(imageData.count / 1024)KB")

        return imageData.base64EncodedString()
    }

    private func resizeImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size

        // Check if resize needed
        guard size.width > maxDimension || size.height > maxDimension else {
            return image
        }

        // Calculate new size maintaining aspect ratio
        let ratio = size.width / size.height
        let newSize: CGSize

        if size.width > size.height {
            newSize = CGSize(width: maxDimension, height: maxDimension / ratio)
        } else {
            newSize = CGSize(width: maxDimension * ratio, height: maxDimension)
        }

        // Resize
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        defer { UIGraphicsEndImageContext() }

        image.draw(in: CGRect(origin: .zero, size: newSize))

        return UIGraphicsGetImageFromCurrentImageContext() ?? image
    }
}

// MARK: - Mock Service for Testing
#if DEBUG
actor MockAIService {
    func generateQuiz(from text: String, subject: String? = nil) async throws -> Quiz {
        // Simulate API delay
        try await Task.sleep(nanoseconds: 2_000_000_000) // 2 seconds

        // Return sample quiz
        return Quiz(
            homeworkId: UUID(),
            questions: Question.sampleList
        )
    }

    func generateQuiz(from image: UIImage, subject: String? = nil) async throws -> Quiz {
        // Simulate API delay for vision processing
        try await Task.sleep(nanoseconds: 2_000_000_000) // 2 seconds

        print("🧪 Mock: Generating quiz from image")

        // Return sample quiz
        return Quiz(
            homeworkId: UUID(),
            questions: Question.sampleList
        )
    }
}
#endif

File: cheater-web/services/aiService.ts
/**
 * AIService - Claude Vision API Integration (Web Version)
 * Simplified for Next.js web environment
 */

import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import {
  Quiz,
  ClaudeResponse,
  convertClaudeResponseToQuiz,
  AIError,
  AIServiceError
} from '../types';

export class AIService {
  private apiKey: string;
  private apiURL = 'https://api.anthropic.com/v1/messages';
  private model = 'claude-sonnet-4-5-20250929';
  private maxTokens = 4096;
  private timeout = 60000; // 60 seconds

  constructor(apiKey?: string) {
    this.apiKey = apiKey || process.env.NEXT_PUBLIC_ANTHROPIC_API_KEY || '';

    // Debug logging
    if (this.apiKey) {
      console.log('✅ API Key loaded:', this.apiKey.substring(0, 10) + '...');
    } else {
      console.error('❌ API Key missing! Check .env.local file.');
    }
  }

  /**
   * Generate quiz from image using Claude Vision API (via Next.js API route)
   */
  async generateQuiz(
    imageUri: string
  ): Promise<{ quiz: Quiz; title: string; subject?: string }> {
    console.log('🚀 Starting quiz generation...');

    try {
      // 1. Process image to base64
      console.log('📸 Processing image...');
      const base64Image = await this.convertImageToBase64(imageUri);

      // 2. Call Next.js API route (server-side proxy to avoid CORS)
      console.log('📤 Sending request to API route...');
      const response = await axios.post(
        '/api/generate-quiz',
        {
          imageBase64: base64Image
        },
        {
          timeout: this.timeout
        }
      );

      console.log('✅ Received response from API');

      // 3. Parse response (includes quiz + title + subject)
      const result = this.parseQuizResponse(response.data);
      console.log('✨ Quiz generated successfully');
      console.log(`📝 Title: ${result.title}`);
      console.log(`📚 Subject: ${result.subject}`);

      return result;
    } catch (error) {
      throw this.handleAPIError(error);
    }
  }

  /**
   * Convert image to base64 using Canvas API
   */
  private async convertImageToBase64(imageUri: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const img = new Image();
      img.crossOrigin = 'anonymous';

      img.onload = () => {
        try {
          // Calculate new dimensions (max 1280px)
          let width = img.width;
          let height = img.height;
          const maxDim = 1280;

          if (width > maxDim || height > maxDim) {
            if (width > height) {
              height = Math.round((height * maxDim) / width);
              width = maxDim;
            } else {
              width = Math.round((width * maxDim) / height);
              height = maxDim;
            }
          }

          // Create canvas and resize
          const canvas = document.createElement('canvas');
          canvas.width = width;
          canvas.height = height;
          const ctx = canvas.getContext('2d');

          if (!ctx) {
            throw new Error('Could not get canvas context');
          }

          ctx.drawImage(img, 0, 0, width, height);

          // Convert to JPEG base64 (0.75 quality)
          const dataUrl = canvas.toDataURL('image/jpeg', 0.75);

          // Extract base64 data (remove "data:image/jpeg;base64," prefix)
          const base64 = dataUrl.split(',')[1];

          // Log size for monitoring
          const sizeKB = Math.round((base64.length * 3) / 4 / 1024);
          console.log(`📸 Image size: ${sizeKB}KB (${width}x${height})`);

          resolve(base64);
        } catch (error) {
          reject(error);
        }
      };

      img.onerror = () => {
        reject(new Error('Failed to load image'));
      };

      img.src = imageUri;
    });
  }

  /**
   * Build prompt for quiz generation
   */
  private buildPrompt(subject?: string): string {
    return `You are an expert educational AI that generates high-quality quiz questions from homework assignments.

Analyze this homework image and generate exactly 10 quiz questions.

${subject ? `The subject is: ${subject}` : 'Auto-detect the subject.'}

Return ONLY a JSON object with this exact structure:
{
  "topic": "maths" | "english" | "science" | "history" | "generic",
  "subtopic": "specific topic (e.g., algebra, grammar)",
  "confidence": 0.0-1.0,
  "questions": [
    {
      "type": "mcq" | "fillBlank" | "shortAnswer",
      "question": "Question text",
      "options": ["A", "B", "C", "D"],  // Only for MCQ
      "correctIndex": 0-3,  // Only for MCQ
      "correctAnswer": "answer",  // For fillBlank and shortAnswer
      "explanation": "Brief explanation"
    }
  ]
}

Requirements:
- Exactly 10 questions
- Mix of question types: ~60% MCQ, ~30% fill-blank, ~10% short-answer
- MCQ must have exactly 4 options
- All questions must have explanations
- Questions should test understanding, not just memorization`;
  }

  /**
   * Parse quiz response from Claude API
   */
  private parseQuizResponse(apiResponse: any): { quiz: Quiz; title: string; subject?: string } {
    try {
      // 1. Extract text from content array
      if (!apiResponse.content || !Array.isArray(apiResponse.content)) {
        throw new Error('Invalid API response structure');
      }

      const textContent = apiResponse.content.find((c: any) => c.type === 'text');
      if (!textContent || !textContent.text) {
        throw new Error('No text content in API response');
      }

      const text = textContent.text;

      // 2. Remove markdown code blocks if present
      const cleanedText = text
        .replace(/```json\n?/g, '')
        .replace(/```\n?/g, '')
        .trim();

      // 3. Find JSON object boundaries
      const jsonStart = cleanedText.indexOf('{');
      const jsonEnd = cleanedText.lastIndexOf('}') + 1;

      if (jsonStart === -1 || jsonEnd === 0) {
        console.error('❌ No JSON object found in response');
        throw new Error('No JSON object found in response');
      }

      const jsonString = cleanedText.substring(jsonStart, jsonEnd);

      // 4. Parse JSON
      const parsed: any = JSON.parse(jsonString);

      // 5. Validate structure
      if (!parsed.questions || parsed.questions.length !== 10) {
        throw new Error(
          `Expected 10 questions, got ${parsed.questions?.length || 0}`
        );
      }

      // 6. Extract title and subject
      const title = parsed.title || 'Homework Assignment';
      const subject = parsed.subject;

      // 7. Convert to Quiz model
      const quiz = convertClaudeResponseToQuiz(parsed, ''); // homeworkId set by caller

      console.log(`✅ Quiz validated: ${quiz.questions.length} questions`);
      console.log(`📚 Topic: ${quiz.topic || 'generic'}`);

      return { quiz, title, subject };
    } catch (error) {
      console.error('❌ Parsing error:', error);
      throw new AIServiceError(
        AIError.ParsingError,
        error instanceof Error ? error.message : 'Failed to parse quiz response'
      );
    }
  }

  /**
   * Handle API errors
   */
  private handleAPIError(error: any): AIServiceError {
    // Log the full error for debugging
    console.error('❌ API Error Details:', {
      code: error.code,
      message: error.message,
      response: error.response?.data,
      status: error.response?.status
    });

    // Timeout
    if (error.code === 'ECONNABORTED' || error.message?.includes('timeout')) {
      return new AIServiceError(
        AIError.Timeout,
        'Request timed out after 60 seconds'
      );
    }

    // Network errors
    if (error.code === 'ENOTFOUND' || error.code === 'ECONNREFUSED') {
      return new AIServiceError(
        AIError.NetworkError,
        'Could not connect to API. Check your internet connection.'
      );
    }

    // HTTP errors
    if (error.response) {
      const status = error.response.status;
      const errorData = error.response.data;

      switch (status) {
        case 401:
          return new AIServiceError(
            AIError.MissingAPIKey,
            `Invalid or missing API key. ${errorData?.error?.message || ''}`,
            401
          );
        case 429:
          return new AIServiceError(
            AIError.QuotaExceeded,
            'Rate limit exceeded. Please wait and try again.',
            429
          );
        case 500:
        case 502:
        case 503:
        case 504:
          return new AIServiceError(
            AIError.ServerError,
            `Server error (${status}). Please try again later.`,
            status
          );
        default:
          return new AIServiceError(
            AIError.InvalidResponse,
            `Unexpected status code: ${status}. ${errorData?.error?.message || ''}`,
            status
          );
      }
    }

    // AIServiceError passthrough
    if (error instanceof AIServiceError) {
      return error;
    }

    // Unknown errors
    return new AIServiceError(
      AIError.NetworkError,
      error.message || 'Unknown error occurred'
    );
  }
}

// Export singleton instance
export const aiService = new AIService();
